extensions/guacamole-auth-cas/src/main/java/org/apache/guacamole/auth/cas/ticket/TicketValidationService.java

```diff
@@ -122,16 +134,44 @@
             }
             
             // Convert remaining attributes that have values to Strings
+            String groupAttribute = confService.getGroupAttribute();
+            // Use cas-member-attribute to retrieve and set group memberships
+            String groupDnFormat = confService.getGroupDnFormat();
+            String groupTemplate = "";
+            if (groupDnFormat != null) {
+                // if CAS is backended to LDAP, groups come in as RFC4514 DN
+                // syntax.  If cas-group-dn-format is set, this strips an
+                // entry such as "CN=Foo,OU=Bar,DC=example,DC=com" to "Foo"
+                groupTemplate = groupDnFormat.replace("%s","([A-Za-z0-9_\\(\\)\\-\\.\\s+]+)");
+                // the underlying parser aggregates all instances of the same
+                // attribute, so we need to be able to parse them out
+                groupTemplate=groupTemplate+",*\\s*";
+            }
+            else {
+                groupTemplate = "([A-Za-z0-9_\\(\\)\\-\\.\\s+]+,*\\s*)";
+            }
+            Pattern pattern = Pattern.compile(groupTemplate);
+
             for (Entry <String, Object> attr : ticketAttrs.entrySet()) {
                 String tokenName = TokenName.canonicalize(attr.getKey(),
                         CAS_ATTRIBUTE_TOKEN_PREFIX);
                 Object value = attr.getValue();
-                if (value != null)
-                    tokens.put(tokenName, value.toString());
+                if (value != null) {
+                    String attrValue = value.toString();
+                    tokens.put(tokenName, attrValue);
+                    if (attr.getKey().equals(groupAttribute)) {
+                        Matcher matcher =
+                            pattern.matcher(attrValue.substring(1,attrValue.length()-1));
+                        while (matcher.find()) {
+                            effectiveGroups.add(matcher.group(1));
+                        }
+                    }
```

Comment:
I believe it was mentioned before that the format of the `memberOf` CAS attribute was `['GROUP1', 'GROUP2', ...]`. Do we know:

* Is this _truly_ the case? Or is it the result of `value.toString()` on some type that could be iterated cleanly? Is there any way to know whether the CAS client implementation returns an actual iterable type?
* If it does indeed provide only a string, where is that format dictated? How does it handle the possible presence of meaningful characters like `[`, `,`, or `]`?

Reply:
Is this truly the case?
=> no, the toString() is converting the collection to something displayable, not iterable. This is the CAS java client implementation. instanceof Collection is the way to test if CAS attribute is multivalued.
In the CAS 2.0 protocol, every value of attribute is separated in a xml tag.

This should look like this (not tested):
```
if (attr.getKey().equals(groupAttribute)) {
  if (attr.getValue() instanceof Collection) {
    for (final Object value : (Collection) attr.getValue()) { 
        effectiveGroups.add(value.toString()); } 
    } else { 
        effectiveGroups.add(attr.getValue().toString()): 
    }
  }
}
```
(wihout dn strip logic)
IMHO